### PR TITLE
add short circuit in BinaryExpr

### LINKED ISSUE #11212: `BinaryExpr` evaluate lacks optimization for `Or` and `And` scenarios
### Describe the bug

As shown in the [code link](https://docs.rs/datafusion-physical-expr/39.0.0/src/datafusion_physical_expr/expressions/binary.rs.html#257), `BinaryExpr` directly calculates the `left` and `right` operands without considering the possibility that `left` might already be `false` and `op` is `And`, or that `left` might be `true` and `op` is `Or`.

For example, in this complex filtering statement, unnecessary calculations will be performed:
```sql
SELECT 
    customer_id,
    first_name,
    last_name,
    email,
    total_purchases
FROM 
    customers
WHERE 
    1 = 0  -- This condition is always false
    AND (
        (total_purchases > 1000 AND EXTRACT(YEAR FROM registration_date) > 2018)
        OR (LOWER(email) LIKE '%@gmail.com' AND loyalty_points > 500)
        OR (
            LENGTH(first_name) + LENGTH(last_name) > 15
            AND SUBSTR(last_name, 1, 1) IN ('M', 'N', 'O', 'P')
            AND DATE_PART('month', last_purchase_date) BETWEEN 6 AND 8
        )
    )
    AND (
        customer_id % 3 = 0
        OR EXISTS (
            SELECT 1 
            FROM orders 
            WHERE orders.customer_id = customers.customer_id
            AND orders.order_status = 'Completed'
            AND orders.order_total > 250
        )
    );
```

### To Reproduce

When evaluating AND or OR expressions

### Expected behavior

When an And expression is executed, a false occurs to stop the computation of the result. 
When an Or expression is executed, a true occurs and the result is stopped.

### Additional context

No

### PR DESCRIPTION
**Pull request by acking-you**
## Which issue does this PR close?

Closes #11212

### CODE REVIEW

File: datafusion/physical-expr/src/expressions/binary.rs [hunk 1]
```diff
@@ -257,7 +257,56 @@ impl PhysicalExpr for BinaryExpr {
...
 
+        #[inline]
+        fn check_short_circuit(arg: &ColumnarValue, op: &Operator) -> bool {
+            let data_type = arg.data_type();
+            if data_type == DataType::Boolean {
+                if *op == Operator::And {
+                    match arg {
+                        ColumnarValue::Array(array) => {
+                            if let Ok(array) = as_boolean_array(&array) {
+                                if array.true_count() == 0 {
```
Comment (dharanad) [inline datafusion/physical-expr/src/expressions/binary.rs:268]: Nit: We can return inline `return array.true_count() == 0` here. Same for other places below

File: datafusion/physical-expr/src/expressions/binary.rs [hunk 2]
```diff
@@ -257,7 +257,56 @@ impl PhysicalExpr for BinaryExpr {
     fn evaluate(&self, batch: &RecordBatch) -> Result<ColumnarValue> {
         use arrow::compute::kernels::numeric::*;
 
+        #[inline]
```
Comment (Dandandan) [inline datafusion/physical-expr/src/expressions/binary.rs:260]: Does this really need `inline`?

File: datafusion/physical-expr/src/expressions/binary.rs [hunk 3]
```diff
@@ -257,7 +257,56 @@ impl PhysicalExpr for BinaryExpr {
...
+        fn check_short_circuit(arg: &ColumnarValue, op: &Operator) -> bool {
+            let data_type = arg.data_type();
+            if data_type == DataType::Boolean {
```
Comment (Dandandan) [inline datafusion/physical-expr/src/expressions/binary.rs:263]: I think this read easier by using `match (arg, *op)` in combination with `if` guard.

```
match (arg, *op) {
    (ColumnarValue::Array(array), Operator::And) => {
       ...
}
```

**Review by alamb — COMMENTED** [review]
Thank you for this contribution @acking-you  -- I am also running some benchmarks on this branch and will report back. This PR looks quite close to me

I also think it would be good to move this optimization upstream into the arrow-rs binary kernels as well

https://docs.rs/arrow/latest/arrow/compute/kernels/boolean/index.html

I will file a ticket to do so later if no one beats me to it

File: datafusion/physical-expr/src/expressions/binary.rs (hunk 2 above)
Comment (jayzhan211) [reply datafusion/physical-expr/src/expressions/binary.rs:260]: I think we don't need `inline`

File: datafusion/physical-expr/src/expressions/binary.rs (hunk 3 above)
Comment (acking-you) [reply datafusion/physical-expr/src/expressions/binary.rs:263]: ok,I'll fix it, thanks for the review.

File: datafusion/physical-expr/src/expressions/binary.rs (hunk 2 above)
Comment (acking-you) [reply datafusion/physical-expr/src/expressions/binary.rs:260]: ok,I'll change this

**Comment by Dandandan** [comment]
> Thank you for this contribution @acking-you -- I am also running some benchmarks on this branch and will report back. This PR looks quite close to me
> 
> I also think it would be good to move this optimization upstream into the arrow-rs binary kernels as well
> 
> https://docs.rs/arrow/latest/arrow/compute/kernels/boolean/index.html
> 
> I will file a ticket to do so later if no one beats me to it

As @jayzhan211 pointed out  https://github.com/apache/datafusion/issues/11212#issuecomment-2205933735 the optimization here is a bit more "advanced" than adding it to a kernel, as we would avoid evaluating the right side altogether (e.g. `ALL_FALSE AND (regexp_like(...))` stops after executing ALL_FALSE)

**Comment by Dandandan** [comment]
I filed https://github.com/apache/datafusion/issues/11262 to make the most out of this optimization in DataFusion.

**Comment by acking-you** [comment]
> > Thank you for this contribution @acking-you -- I am also running some benchmarks on this branch and will report back. This PR looks quite close to me
> > I also think it would be good to move this optimization upstream into the arrow-rs binary kernels as well
> > https://docs.rs/arrow/latest/arrow/compute/kernels/boolean/index.html
> > I will file a ticket to do so later if no one beats me to it
> 
> As @jayzhan211 pointed out [#11212 (comment)](https://github.com/apache/datafusion/issues/11212#issuecomment-2205933735) the optimization here is a bit more "advanced" than adding it to a kernel, as we would avoid evaluating the right side altogether (e.g. `ALL_FALSE AND (regexp_like(...))` stops after executing ALL_FALSE whereas changing the kernel still would execute the `regexp_like`)

That's exactly what I was getting at.
Unless datafusion significantly overhauls the calculation logic here, it's unlikely a solution could be achieved by modifying the kernel's code.

**Comment by Dandandan** [comment]
These are my bench results 🤔 

```
--------------------
Benchmark tpch_mem_sf1.json
--------------------
┏━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃     main ┃ add_short_circuit ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 1     │  77.06ms │           79.34ms │     no change │
│ QQuery 2     │  16.48ms │           15.84ms │     no change │
│ QQuery 3     │  23.68ms │           25.64ms │  1.08x slower │
│ QQuery 4     │  21.89ms │           21.47ms │     no change │
│ QQuery 5     │  35.82ms │           39.28ms │  1.10x slower │
│ QQuery 6     │   5.67ms │            6.42ms │  1.13x slower │
│ QQuery 7     │  82.14ms │           79.52ms │     no change │
│ QQuery 8     │  20.69ms │           20.54ms │     no change │
│ QQuery 9     │  45.88ms │           46.78ms │     no change │
│ QQuery 10    │  42.23ms │           43.26ms │     no change │
│ QQuery 11    │  32.57ms │           33.04ms │     no change │
│ QQuery 12    │  23.49ms │           22.80ms │     no change │
│ QQuery 13    │  29.71ms │           28.56ms │     no change │
│ QQuery 14    │   7.98ms │            7.28ms │ +1.10x faster │
│ QQuery 15    │  14.45ms │           14.54ms │     no change │
│ QQuery 16    │  18.54ms │           17.73ms │     no change │
│ QQuery 17    │  61.68ms │           69.53ms │  1.13x slower │
│ QQuery 18    │ 131.52ms │          141.04ms │  1.07x slower │
│ QQuery 19    │  22.75ms │           22.61ms │     no change │
│ QQuery 20    │  32.10ms │           31.86ms │     no change │
│ QQuery 21    │  98.63ms │           95.29ms │     no change │
│ QQuery 22    │   9.83ms │            9.53ms │     no change │
└──────────────┴──────────┴───────────────────┴───────────────┘
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┓
┃ Benchmark Summary                ┃          ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━┩
│ Total Time (main)                │ 854.79ms │
│ Total Time (add_short_circuit)   │ 871.88ms │
│ Average Time (main)              │  38.85ms │
│ Average Time (add_short_circuit) │  39.63ms │
│ Queries Faster                   │        1 │
│ Queries Slower                   │        5 │
│ Queries with No Change           │       16 │
└──────────────────────────────────┴──────────┘
--------------------
Benchmark tpch_sf1.json
--------------------
┏━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━┓
┃ Query        ┃     main ┃ add_short_circuit ┃       Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━┩
│ QQuery 1     │ 148.26ms │          148.95ms │    no change │
│ QQuery 2     │  67.87ms │           66.55ms │    no change │
│ QQuery 3     │  69.60ms │           69.37ms │    no change │
│ QQuery 4     │  52.01ms │           51.70ms │    no change │
│ QQuery 5     │  82.18ms │           80.46ms │    no change │
│ QQuery 6     │  31.38ms │           74.62ms │ 2.38x slower │
│ QQuery 7     │ 105.18ms │          102.44ms │    no change │
│ QQuery 8     │  93.49ms │           92.66ms │    no change │
│ QQuery 9     │ 134.92ms │          135.27ms │    no change │
│ QQuery 10    │ 125.29ms │          122.37ms │    no change │
│ QQuery 11    │  48.71ms │           48.00ms │    no change │
│ QQuery 12    │  82.64ms │           81.72ms │    no change │
│ QQuery 13    │ 191.46ms │          189.70ms │    no change │
│ QQuery 14    │  49.91ms │           48.30ms │    no change │
│ QQuery 15    │  63.63ms │           63.35ms │    no change │
│ QQuery 16    │  53.03ms │           51.24ms │    no change │
│ QQuery 17    │ 128.33ms │          127.72ms │    no change │
│ QQuery 18    │ 164.81ms │          167.68ms │    no change │
│ QQuery 19    │  89.80ms │           89.78ms │    no change │
│ QQuery 20    │  81.85ms │           84.25ms │    no change │
│ QQuery 21    │ 128.79ms │          131.90ms │    no change │
│ QQuery 22    │  40.00ms │           39.45ms │    no change │
└──────────────┴──────────┴───────────────────┴──────────────┘
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━┓
┃ Benchmark Summary                ┃           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━┩
│ Total Time (main)                │ 2033.15ms │
│ Total Time (add_short_circuit)   │ 2067.50ms │
│ Average Time (main)              │   92.42ms │
│ Average Time (add_short_circuit) │   93.98ms │
│ Queries Faster                   │         0 │
│ Queries Slower                   │         1 │
│ Queries with No Change           │        21 │
└──────────────────────────────────┴───────────┘
```

**Comment by Dandandan** [comment]
I merged against latest `main`, it seems in last 2 days something yielded a >2x speedup for query 6.

I get now these mixed results on my (Apple m1 pro) machine:
```
--------------------
Benchmark tpch_mem_sf1.json
--------------------
┏━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃     main ┃ add_short_circuit ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 1     │  77.06ms │           78.03ms │     no change │
│ QQuery 2     │  16.48ms │           16.45ms │     no change │
│ QQuery 3     │  23.68ms │           26.24ms │  1.11x slower │
│ QQuery 4     │  21.89ms │           22.47ms │     no change │
│ QQuery 5     │  35.82ms │           38.72ms │  1.08x slower │
│ QQuery 6     │   5.67ms │            6.98ms │  1.23x slower │
│ QQuery 7     │  82.14ms │           76.61ms │ +1.07x faster │
│ QQuery 8     │  20.69ms │           21.64ms │     no change │
│ QQuery 9     │  45.88ms │           48.10ms │     no change │
│ QQuery 10    │  42.23ms │           44.08ms │     no change │
│ QQuery 11    │  32.57ms │           32.57ms │     no change │
│ QQuery 12    │  23.49ms │           23.07ms │     no change │
│ QQuery 13    │  29.71ms │           28.15ms │ +1.06x faster │
│ QQuery 14    │   7.98ms │            7.16ms │ +1.11x faster │
│ QQuery 15    │  14.45ms │           14.77ms │     no change │
│ QQuery 16    │  18.54ms │           19.55ms │  1.05x slower │
│ QQuery 17    │  61.68ms │           63.54ms │     no change │
│ QQuery 18    │ 131.52ms │          146.46ms │  1.11x slower │
│ QQuery 19    │  22.75ms │           22.64ms │     no change │
│ QQuery 20    │  32.10ms │           30.86ms │     no change │
│ QQuery 21    │  98.63ms │           97.17ms │     no change │
│ QQuery 22    │   9.83ms │           10.55ms │  1.07x slower │
└──────────────┴──────────┴───────────────────┴───────────────┘
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┓
┃ Benchmark Summary                ┃          ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━┩
│ Total Time (main)                │ 854.79ms │
│ Total Time (add_short_circuit)   │ 875.83ms │
│ Average Time (main)              │  38.85ms │
│ Average Time (add_short_circuit) │  39.81ms │
│ Queries Faster                   │        3 │
│ Queries Slower                   │        6 │
│ Queries with No Change           │       13 │
└──────────────────────────────────┴──────────┘
--------------------
Benchmark tpch_sf1.json
--------------------
┏━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━┓
┃ Query        ┃     main ┃ add_short_circuit ┃    Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━┩
│ QQuery 1     │ 148.26ms │          148.34ms │ no change │
│ QQuery 2     │  67.87ms │           67.88ms │ no change │
│ QQuery 3     │  69.60ms │           71.52ms │ no change │
│ QQuery 4     │  52.01ms │           53.10ms │ no change │
│ QQuery 5     │  82.18ms │           83.01ms │ no change │
│ QQuery 6     │  31.38ms │           32.28ms │ no change │
│ QQuery 7     │ 105.18ms │          102.67ms │ no change │
│ QQuery 8     │  93.49ms │           92.19ms │ no change │
│ QQuery 9     │ 134.92ms │          136.07ms │ no change │
│ QQuery 10    │ 125.29ms │          124.37ms │ no change │
│ QQuery 11    │  48.71ms │           49.17ms │ no change │
│ QQuery 12    │  82.64ms │           82.86ms │ no change │
│ QQuery 13    │ 191.46ms │          190.90ms │ no change │
│ QQuery 14    │  49.91ms │           48.91ms │ no change │
│ QQuery 15    │  63.63ms │           61.70ms │ no change │
│ QQuery 16    │  53.03ms │           52.23ms │ no change │
│ QQuery 17    │ 128.33ms │          128.75ms │ no change │
│ QQuery 18    │ 164.81ms │          167.84ms │ no change │
│ QQuery 19    │  89.80ms │           88.97ms │ no change │
│ QQuery 20    │  81.85ms │           83.98ms │ no change │
│ QQuery 21    │ 128.79ms │          132.33ms │ no change │
│ QQuery 22    │  40.00ms │           40.72ms │ no change │
└──────────────┴──────────┴───────────────────┴───────────┘
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━┓
┃ Benchmark Summary                ┃           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━┩
│ Total Time (main)                │ 2033.15ms │
│ Total Time (add_short_circuit)   │ 2039.79ms │
│ Average Time (main)              │   92.42ms │
│ Average Time (add_short_circuit) │   92.72ms │
│ Queries Faster                   │         0 │
│ Queries Slower                   │         0 │
│ Queries with No Change           │        22 │
└──────────────────────────────────┴───────────┘
```

**Comment by jayzhan211** [comment]
```
Benchmark tpch_sf10.json
--------------------
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃      main ┃ add_short_circuit ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 1     │ 1016.92ms │          818.68ms │ +1.24x faster │
│ QQuery 2     │  146.80ms │          131.71ms │ +1.11x faster │
│ QQuery 3     │  445.80ms │          417.92ms │ +1.07x faster │
│ QQuery 4     │  228.00ms │          212.54ms │ +1.07x faster │
│ QQuery 5     │  616.29ms │          562.94ms │ +1.09x faster │
│ QQuery 6     │  193.28ms │          188.68ms │     no change │
│ QQuery 7     │  806.95ms │          853.22ms │  1.06x slower │
│ QQuery 8     │  582.18ms │          636.72ms │  1.09x slower │
│ QQuery 9     │  926.48ms │          974.76ms │  1.05x slower │
│ QQuery 10    │  632.74ms │          621.95ms │     no change │
│ QQuery 11    │   83.91ms │           80.06ms │     no change │
│ QQuery 12    │  318.26ms │          322.94ms │     no change │
│ QQuery 13    │  495.16ms │          508.65ms │     no change │
│ QQuery 14    │  261.28ms │          271.91ms │     no change │
│ QQuery 15    │  386.88ms │          392.05ms │     no change │
│ QQuery 16    │  124.54ms │          121.21ms │     no change │
│ QQuery 17    │ 1301.28ms │         1304.97ms │     no change │
│ QQuery 18    │ 1356.90ms │         1364.60ms │     no change │
│ QQuery 19    │  509.38ms │          523.05ms │     no change │
│ QQuery 20    │  504.72ms │          501.61ms │     no change │
│ QQuery 21    │ 1033.29ms │         1041.01ms │     no change │
│ QQuery 22    │  136.02ms │          134.02ms │     no change │
└──────────────┴───────────┴───────────────────┴───────────────┘
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┓
┃ Benchmark Summary                ┃            ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━┩
│ Total Time (main)                │ 12107.07ms │
│ Total Time (add_short_circuit)   │ 11985.18ms │
│ Average Time (main)              │   550.32ms │
│ Average Time (add_short_circuit) │   544.78ms │
│ Queries Faster                   │          5 │
│ Queries Slower                   │          3 │
│ Queries with No Change           │         14 │
└──────────────────────────────────┴────────────┘
```

I'm quite surprise that the change is slower for some queries

**Comment by jayzhan211** [comment]
And this
```
Benchmark tpch_mem_sf10.json
--------------------
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃      main ┃ add_short_circuit ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 1     │  669.46ms │          667.62ms │     no change │
│ QQuery 2     │  106.48ms │          106.80ms │     no change │
│ QQuery 3     │  199.31ms │          213.33ms │  1.07x slower │
│ QQuery 4     │  179.86ms │          188.93ms │  1.05x slower │
│ QQuery 5     │  461.28ms │          416.30ms │ +1.11x faster │
│ QQuery 6     │   35.52ms │           38.11ms │  1.07x slower │
│ QQuery 7     │  813.69ms │          824.49ms │     no change │
│ QQuery 8     │  271.07ms │          266.10ms │     no change │
│ QQuery 9     │  650.23ms │          649.26ms │     no change │
│ QQuery 10    │  354.09ms │          357.90ms │     no change │
│ QQuery 11    │  338.08ms │          348.68ms │     no change │
│ QQuery 12    │  160.68ms │          165.02ms │     no change │
│ QQuery 13    │  338.63ms │          312.28ms │ +1.08x faster │
│ QQuery 14    │   49.03ms │           51.79ms │  1.06x slower │
│ QQuery 15    │  131.30ms │          109.20ms │ +1.20x faster │
│ QQuery 16    │  106.34ms │          119.37ms │  1.12x slower │
│ QQuery 17    │ 1049.16ms │          986.42ms │ +1.06x faster │
│ QQuery 18    │ 3475.32ms │         3814.02ms │  1.10x slower │
│ QQuery 19    │  135.61ms │          135.27ms │     no change │
│ QQuery 20    │  339.62ms │          382.27ms │  1.13x slower │
│ QQuery 21    │ 1062.58ms │         1126.34ms │  1.06x slower │
│ QQuery 22    │   72.90ms │           77.33ms │  1.06x slower │
└──────────────┴───────────┴───────────────────┴───────────────┘
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━┓
┃ Benchmark Summary                ┃            ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━┩
│ Total Time (main)                │ 11000.25ms │
│ Total Time (add_short_circuit)   │ 11356.82ms │
│ Average Time (main)              │   500.01ms │
│ Average Time (add_short_circuit) │   516.22ms │
│ Queries Faster                   │          4 │
│ Queries Slower                   │          9 │
│ Queries with No Change           │          9 │
└──────────────────────────────────┴────────────┘
```

**Comment by acking-you** [comment]
> I merged against latest `main`, it seems in last 2 days something yielded a >2x speedup for query 6.
> 
> I get now these mixed results on my (Apple m1 pro) machine:
> 
> ```
> --------------------
> Benchmark tpch_mem_sf1.json
> --------------------
> ┏━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
> ┃ Query        ┃     main ┃ add_short_circuit ┃        Change ┃
> ┡━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
> │ QQuery 1     │  77.06ms │           78.03ms │     no change │
> │ QQuery 2     │  16.48ms │           16.45ms │     no change │
> │ QQuery 3     │  23.68ms │           26.24ms │  1.11x slower │
> │ QQuery 4     │  21.89ms │           22.47ms │     no change │
> │ QQuery 5     │  35.82ms │           38.72ms │  1.08x slower │
> │ QQuery 6     │   5.67ms │            6.98ms │  1.23x slower │
> │ QQuery 7     │  82.14ms │           76.61ms │ +1.07x faster │
> │ QQuery 8     │  20.69ms │           21.64ms │     no change │
> │ QQuery 9     │  45.88ms │           48.10ms │     no change │
> │ QQuery 10    │  42.23ms │           44.08ms │     no change │
> │ QQuery 11    │  32.57ms │           32.57ms │     no change │
> │ QQuery 12    │  23.49ms │           23.07ms │     no change │
> │ QQuery 13    │  29.71ms │           28.15ms │ +1.06x faster │
> │ QQuery 14    │   7.98ms │            7.16ms │ +1.11x faster │
> │ QQuery 15    │  14.45ms │           14.77ms │     no change │
> │ QQuery 16    │  18.54ms │           19.55ms │  1.05x slower │
> │ QQuery 17    │  61.68ms │           63.54ms │     no change │
> │ QQuery 18    │ 131.52ms │          146.46ms │  1.11x slower │
> │ QQuery 19    │  22.75ms │           22.64ms │     no change │
> │ QQuery 20    │  32.10ms │           30.86ms │     no change │
> │ QQuery 21    │  98.63ms │           97.17ms │     no change │
> │ QQuery 22    │   9.83ms │           10.55ms │  1.07x slower │
> └──────────────┴──────────┴───────────────────┴───────────────┘
> ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━┓
> ┃ Benchmark Summary                ┃          ┃
> ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━┩
> │ Total Time (main)                │ 854.79ms │
> │ Total Time (add_short_circuit)   │ 875.83ms │
> │ Average Time (main)              │  38.85ms │
> │ Average Time (add_short_circuit) │  39.81ms │
> │ Queries Faster                   │        3 │
> │ Queries Slower                   │        6 │
> │ Queries with No Change           │       13 │
> └──────────────────────────────────┴──────────┘
> --------------------
> Benchmark tpch_sf1.json
> --------------------
> ┏━━━━━━━━━━━━━━┳━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━┓
> ┃ Query        ┃     main ┃ add_short_circuit ┃    Change ┃
> ┡━━━━━━━━━━━━━━╇━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━┩
> │ QQuery 1     │ 148.26ms │          148.34ms │ no change │
> │ QQuery 2     │  67.87ms │           67.88ms │ no change │
> │ QQuery 3     │  69.60ms │           71.52ms │ no change │
> │ QQuery 4     │  52.01ms │           53.10ms │ no change │
> │ QQuery 5     │  82.18ms │           83.01ms │ no change │
> │ QQuery 6     │  31.38ms │           32.28ms │ no change │
> │ QQuery 7     │ 105.18ms │          102.67ms │ no change │
> │ QQuery 8     │  93.49ms │           92.19ms │ no change │
> │ QQuery 9     │ 134.92ms │          136.07ms │ no change │
> │ QQuery 10    │ 125.29ms │          124.37ms │ no change │
> │ QQuery 11    │  48.71ms │           49.17ms │ no change │
> │ QQuery 12    │  82.64ms │           82.86ms │ no change │
> │ QQuery 13    │ 191.46ms │          190.90ms │ no change │
> │ QQuery 14    │  49.91ms │           48.91ms │ no change │
> │ QQuery 15    │  63.63ms │           61.70ms │ no change │
> │ QQuery 16    │  53.03ms │           52.23ms │ no change │
> │ QQuery 17    │ 128.33ms │          128.75ms │ no change │
> │ QQuery 18    │ 164.81ms │          167.84ms │ no change │
> │ QQuery 19    │  89.80ms │           88.97ms │ no change │
> │ QQuery 20    │  81.85ms │           83.98ms │ no change │
> │ QQuery 21    │ 128.79ms │          132.33ms │ no change │
> │ QQuery 22    │  40.00ms │           40.72ms │ no change │
> └──────────────┴──────────┴───────────────────┴───────────┘
> ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━┓
> ┃ Benchmark Summary                ┃           ┃
> ┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━┩
> │ Total Time (main)                │ 2033.15ms │
> │ Total Time (add_short_circuit)   │ 2039.79ms │
> │ Average Time (main)              │   92.42ms │
> │ Average Time (add_short_circuit) │   92.72ms │
> │ Queries Faster                   │         0 │
> │ Queries Slower                   │         0 │
> │ Queries with No Change           │        22 │
> └──────────────────────────────────┴───────────┘
> ```

I retested several times, and the results varied. There wasn't a significant improvement in the TPC-H benchmark scenarios🥲🥲. However, in our internal extreme SQL use cases (where `left` joins mostly hit and the `right` side is computationally heavy), I saw a fourfold performance boost.

File: datafusion/physical-expr/src/expressions/binary.rs [hunk 4]
```diff
@@ -257,7 +257,50 @@ impl PhysicalExpr for BinaryExpr {
...
         use arrow::compute::kernels::numeric::*;
 
+        fn check_short_circuit(arg: &ColumnarValue, op: &Operator) -> bool {
+            let data_type = arg.data_type();
+            match (data_type, op) {
+                (DataType::Boolean, Operator::And) => {
+                    match arg {
+                        ColumnarValue::Array(array) => {
+                            if let Ok(array) = as_boolean_array(&array) {
+                                return array.true_count() == 0;
```
Comment (alamb) [inline datafusion/physical-expr/src/expressions/binary.rs:267]: FWIW this `true_count` operation requires checking all values (which should still be fast, but I wanted to point out it was not free / constant time): https://docs.rs/arrow-array/52.0.0/src/arrow_array/array/boolean_array.rs.html#142

**Review by alamb — COMMENTED** [review]
my benchmark runs had some issues, I am running them again

**Comment by alamb** [comment]
> I merged against latest main, it seems in last 2 days something yielded a >2x speedup for query 6.

Maybe https://github.com/apache/datafusion/pull/11211 🤔

**Comment by alamb** [comment]
Here are my benchmark results

```
--------------------
Benchmark tpch_mem.json
--------------------
┏━━━━━━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Query        ┃ main_base ┃ add_short_circuit ┃        Change ┃
┡━━━━━━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ QQuery 1     │  200.29ms │          209.62ms │     no change │
│ QQuery 2     │   35.62ms │           34.17ms │     no change │
│ QQuery 3     │   60.02ms │           62.15ms │     no change │
│ QQuery 4     │   63.80ms │           63.97ms │     no change │
│ QQuery 5     │   96.94ms │          102.43ms │  1.06x slower │
│ QQuery 6     │   14.70ms │           14.40ms │     no change │
│ QQuery 7     │  199.35ms │          207.52ms │     no change │
│ QQuery 8     │   40.29ms │           40.32ms │     no change │
│ QQuery 9     │  109.85ms │          106.79ms │     no change │
│ QQuery 10    │  103.39ms │          107.30ms │     no change │
│ QQuery 11    │   72.28ms │           74.07ms │     no change │
│ QQuery 12    │   45.57ms │           48.88ms │  1.07x slower │
│ QQuery 13    │   78.70ms │           81.60ms │     no change │
│ QQuery 14    │   18.92ms │           19.50ms │     no change │
│ QQuery 15    │   30.10ms │           30.37ms │     no change │
│ QQuery 16    │   44.68ms │           41.71ms │ +1.07x faster │
│ QQuery 17    │  159.93ms │          162.91ms │     no change │
│ QQuery 18    │  476.45ms │          457.49ms │     no change │
│ QQuery 19    │   47.57ms │           45.01ms │ +1.06x faster │
│ QQuery 20    │  104.00ms │           92.20ms │ +1.13x faster │
│ QQuery 21    │  290.77ms │          284.74ms │     no change │
│ QQuery 22    │   22.52ms │           22.87ms │     no change │
└──────────────┴───────────┴───────────────────┴───────────────┘
┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━┓
┃ Benchmark Summary                ┃           ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━┩
│ Total Time (main_base)           │ 2315.72ms │
│ Total Time (add_short_circuit)   │ 2310.01ms │
│ Average Time (main_base)         │  105.26ms │
│ Average Time (add_short_circuit) │  105.00ms │
│ Queries Faster                   │         3 │
│ Queries Slower                   │         2 │
│ Queries with No Change           │        17 │
└──────────────────────────────────┴───────────┘
```

So similarly I don't see much better or worse

However, I think the end to end benchmarks might not be as good to evaluate this change as some micro benchmarks  -- for example maybe we could extend https://github.com/apache/datafusion/blob/main/datafusion/core/benches/filter_query_sql.rs with the type of predicates this optimization would help 🤔

File: datafusion/physical-expr/src/expressions/binary.rs (hunk 4 above)
Comment (Dandandan) [reply datafusion/physical-expr/src/expressions/binary.rs:267]: We probably can speed it up by providing a  function that doesn't count but provides whether the bitmap is all true / all false 🤔

**Comment by alamb** [comment]
What is the status of this PR? Are we waiting on some microbenchmarks (`cargo bench` style) to show improvement?

**Pull request closed without merging**